src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1582,58 +1718,98 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMReplayDeltaResponse> const& m)
+                ledgerSeq > app_.getLedgerMaster().getValidLedgerIndex() + 10)
+            {
+                return badData(
+                    "Invalid ledger sequence " + std::to_string(ledgerSeq));
+            }
+        }
+    }
+
+    // Verify ledger info type
+    if (m->type() < protocol::liBASE || m->type() > protocol::liTS_CANDIDATE)
+        return badData("Invalid ledger info type");
+
+    // Verify ledger nodes
+    if (m->nodes_size() <= 0 || m->nodes_size() > Tuning::maxReplyNodes)
+    {
+        return badData(
+            "Invalid Ledger/TXset nodes " + std::to_string(m->nodes_size()));
     }
 
+    // Verify reply error
+    if (m->has_error() &&
+        (m->error() < protocol::reNO_LEDGER ||
+         m->error() > protocol::reBAD_REQUEST))
+    {
+        return badData("Invalid reply error");
+    }
+
+    // If there is a request cookie, attempt to relay the message
     if (m->has_requestcookie())
     {
-        std::shared_ptr<Peer> target =
-            overlay_.findPeerByShortID(m->requestcookie());
-        if (target)
+        if (auto peer = overlay_.findPeerByShortID(m->requestcookie()))
         {
             m->clear_requestcookie();
-            target->send(
-                std::make_shared<Message>(packet, protocol::mtLEDGER_DATA));
+            peer->send(std::make_shared<Message>(*m, protocol::mtLEDGER_DATA));
         }
         else
         {
             JLOG(p_journal_.info()) << "Unable to route TX/ledger data reply";
-            fee_ = Resource::feeUnwantedData;
         }
         return;
     }
 
-    if (!stringIsUint256Sized(m->ledgerhash()))
-    {
-        JLOG(p_journal_.warn()) << "TX candidate reply with invalid hash size";
-        fee_ = Resource::feeInvalidRequest;
-        return;
-    }
-
-    uint256 const hash{m->ledgerhash()};
+    uint256 const ledgerHash{m->ledgerhash()};
 
+    // Otherwise check if received data for a candidate transaction set
     if (m->type() == protocol::liTS_CANDIDATE)
     {
-        // got data for a candidate transaction set
-        std::weak_ptr<PeerImp> weak = shared_from_this();
+        std::weak_ptr<PeerImp> weak{shared_from_this()};
         app_.getJobQueue().addJob(
-            jtTXN_DATA, "recvPeerData", [weak, hash, m](Job&) {
+            jtTXN_DATA, "recvPeerData", [weak, ledgerHash, m](Job&) {
                 if (auto peer = weak.lock())
-                    peer->app_.getInboundTransactions().gotData(hash, peer, m);
+                {
+                    peer->app_.getInboundTransactions().gotData(
+                        ledgerHash, peer, m);
+                }
             });
         return;
     }
 
-    if (!app_.getInboundLedgers().gotLedgerData(hash, shared_from_this(), m))
-    {
-        JLOG(p_journal_.trace()) << "Got data for unwanted ledger";
-        fee_ = Resource::feeUnwantedData;
-    }
+    // Consume the message
+    app_.getInboundLedgers().gotLedgerData(ledgerHash, shared_from_this(), m);
```

Comment:
The old code checks result of `app_.getInboundLedgers().gotLedgerData(ledgerHash, shared_from_this(), m);` and may charge `Resource::feeUnwantedData`. Why is it removed?

Reply:
A return of false means this node is no longer acquiring the ledger but it might still make use of it. The acquiring likely ended because the node received all the ledger data from the peers it requested from. So this case isn't really UnwantedData, the node requested it. Charging the last peer to reply (especially if it was relayed) seems incorrect and can lead to unnecessary disconnects. Please let me know if you think otherwise.

Reply:
make sense. thanks